### PR TITLE
Remove banned C++ features from Rewards codebase

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/34477

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/core/wallet/wallet_balance.cc [hunk 1]
```diff
@@ -22,16 +22,10 @@ WalletBalance::WalletBalance(RewardsEngineImpl& engine) : engine_(engine) {}
...
-                     base::Unretained(this), std::move(callback));
-
   engine_->database()->GetSpendableUnblindedTokensByBatchTypes(
       {mojom::CredsBatchType::PROMOTION},
-      [callback = std::make_shared<decltype(tokens_callback)>(std::move(
-           tokens_callback))](std::vector<mojom::UnblindedTokenPtr> list) {
-        std::move(*callback).Run(std::move(list));
-      });
+      base::BindOnce(&WalletBalance::OnGetUnblindedTokens,
+                     base::Unretained(this), std::move(callback)));
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/wallet/wallet_balance.cc:28]: This is an existing callback (converted from `std::function` to `OnceCallback`) and will be addressed in https://github.com/brave/brave-browser/issues/35805.

File: components/brave_rewards/core/contribution/contribution.cc [hunk 2]
```diff
@@ -530,78 +527,72 @@ void Contribution::CreateNewEntry(const std::string& wallet_type,
...
-      std::make_shared<mojom::ContributionQueuePtr>(std::move(queue));
-
-  auto save_callback = std::bind(&Contribution::OnEntrySaved, this, _1,
-                                 contribution->contribution_id, wallet_type,
-                                 *balance, shared_queue);
-
-  engine_->database()->SaveContributionInfo(contribution->Clone(),
-                                            save_callback);
+  engine_->database()->SaveContributionInfo(
+      contribution->Clone(),
```
Comment (szilardszaloki) [inline components/brave_rewards/core/contribution/contribution.cc:531]: Could be moved if you saved `contribution->contribution_id` to a local variable.

Comment (zenparsing) [reply components/brave_rewards/core/contribution/contribution.cc:531]: I wanted to err on the side of caution with respect to use after move, and I became exhausted at a certain point in trying to reason through and correct every call.

File: components/brave_rewards/core/contribution/contribution_unblinded.cc [hunk 3]
```diff
@@ -233,47 +235,45 @@ void Unblinded::PrepareTokens(
...
+    mojom::Result result) {
   if (result != mojom::Result::OK) {
     engine_->LogError(FROM_HERE) << "Failed to reserve unblinded tokens";
-    callback(mojom::Result::FAILED);
+    std::move(callback).Run(mojom::Result::FAILED);
     return;
   }
 
-  if (!shared_contribution) {
+  if (!contribution) {
```
Comment (szilardszaloki) [inline components/brave_rewards/core/contribution/contribution_unblinded.cc:258]: I noticed this at a few places: such a check was not present previously (would have been `if (!*shared_contribution) { /* ... */ }`). But then I don't think we rely on the side effects anywhere (or if there's any side effects at all).

Comment (zenparsing) [reply components/brave_rewards/core/contribution/contribution_unblinded.cc:258]: It changes the condition to check at one less fewer levels of indirection, which seems fine to me.

Comment (szilardszaloki) [reply components/brave_rewards/core/contribution/contribution_unblinded.cc:258]: 👍🏻

File: components/brave_rewards/core/legacy/media/github.cc [hunk 4]
```diff
@@ -212,32 +209,35 @@ void GitHub::ProcessActivityFromUrl(uint64_t window_id,
...
 
 void GitHub::ProcessMedia(const base::flat_map<std::string, std::string> parts,
                           const mojom::VisitData& visit_data) {
   const std::string user_name = GetUserNameFromURL(visit_data.path);
   const std::string url = GetProfileAPIURL(user_name);
   auto iter = parts.find("duration");
-  uint64_t duration = iter != parts.end() ? std::stoull(iter->second) : 0U;
+  uint64_t duration = 0;
+  if (iter != parts.end()) {
+    base::StringToUint64(iter->second, &duration);
```
Comment (szilardszaloki) [inline components/brave_rewards/core/legacy/media/github.cc:224]: I think we should check the return value here — there's a few other instances as well.

**Review by szilardszaloki — APPROVED** [review]

File: components/brave_rewards/core/legacy/media/github.cc (hunk 4 above)
Comment (zenparsing) [reply components/brave_rewards/core/legacy/media/github.cc:224]: Unfortunately, it's not really clear at some of the call sites what the failure semantics should be. `base::ToString*` will do a best effort conversion, which is probably the safest way to go in this PR.

**Pull request merged**